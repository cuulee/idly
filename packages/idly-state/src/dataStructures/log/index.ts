import { weakCache } from 'idly-common/lib/misc/weakCache';

import {
  iterableFlattenToSet,
  setCreate,
  setFind,
  setIntersect,
  setSome,
} from '../set';

export type ModifiedId = string;

export type Log = ReadonlyArray<Entry>;

export type Entry = ReadonlySet<string>;

export const validVirginId = (str: string) => str.indexOf('#') === 0;

export const validModifiedId = (str: string) => !validVirginId(str);

/**
 * Key Terms
 * log - is the linear history of all modified ids
 * entry - is the smallest revertible set of log i.e the ids in
 *         entry would need to be added or reverted together to be legal modification of log.
 * virginId - is the untouched OSM id, eg n1212, r121
 * modifiedId - is `virginId#version` eg, n121#121, r12#11
 * version - the part after hash # eg, for 12#0 version is 0.
 * currentId - is the most recent modifiedId in the log
 * baseId - the virginId that  relevant id descended from
 */

export const logGetBaseIds: (log: Log) => ReadonlySet<string> = weakCache(
  (log: Log) =>
    log.reduce((prev: Set<string>, cur) => {
      cur.forEach(index => prev.add(baseId(index)));
      return prev;
    }, setCreate())
);

export const logGetCurrentIds: (log: Log) => ReadonlySet<string> = weakCache(
  (log: Log) =>
    new Set(
      log
        .reduce((prev, entry) => {
          entry.forEach(
            index => !prev.has(baseId(index)) && prev.set(baseId(index), index)
          );
          return prev;
        }, new Map<string, string>())
        .values()
    )
);

export const logGetEveryId: (log: Log) => ReadonlySet<string> = weakCache(
  (log: Log) => iterableFlattenToSet<string>(log)
);

/**
 * This function takes a virginId and automatically adds new versions of
 * it in all entries which has atleast one of the provided baseIds.
 * Useful when you have a way which is a parent of modified ID and you want
 * to make it appear this guy always existed.
 * @param virginId - The virginId that you want to exist since starting
 * @param baseIds - The baseIds is the original id which modifiedId descended from
 */
export const logRewrite = (log: Log, virginId: string, baseIds: Set<string>) =>
  log
    .slice(0)
    .reverse()
    .reduce((l, e) => {
      let newEntryArray = [...e];
      if (newEntryArray.map(baseId).find(id => baseIds.has(id))) {
        newEntryArray = [
          ...newEntryArray,
          logGenerateNextModifiedId(virginId)(l),
        ];
      }
      return logAddEntry(setCreate(newEntryArray))(l);
    }, logCreate());

export const logRecreate = (...entries: Entry[]) =>
  entries
    .slice(0)
    .reverse()
    .reduce((l, e) => {
      const newLog = logAddEntry(e)(l);
      if (l === newLog) {
        throw new Error('malformed entries');
      }
      return newLog;
    }, logCreate());

export const logCreate = (): Log => Object.freeze([]);

export const doesLogContain = (entry: Entry) => (log: Log) =>
  log.some(e => setIntersect(entry, e).size > 0);

export const logAddEntry = (entry: Entry) => (log: Log) =>
  setSome(index => doesLogHaveNewerOrCurrentVersion(index)(log), entry)
    ? log
    : Object.freeze([entry, ...log]);

export const doesLogHaveNewerOrCurrentVersion = (index: string) => (log: Log) =>
  logGetLatestVersion(index)(log) >= modifiedIdGetVersion(index);

export const modifiedIdGetVersion = (index: ModifiedId) =>
  parseInt(modifiedIdParse(index)[1], 10);

export const logGetLatestVersion = (id: string) => (log: Log) => {
  const finder = entryFindModifiedId(baseId(id));
  let index;
  for (const entry of log) {
    index = finder(entry);
    if (index) {
      return modifiedIdGetVersion(index);
    }
  }
  return -1;
};

export const logGenerateNextModifiedId = (id: string) => (log: Log) =>
  `${baseId(id)}#${logGetLatestVersion(id)(log) + 1}`;

export const entryFindModifiedId = (indexOrId: string) => (entry: Entry) =>
  setFind(i => baseId(i) === indexOrId, entry);

export const baseId = (index: ModifiedId) => modifiedIdParse(index)[0];

const modifiedIdParse = (index: ModifiedId) => index.split('#');

export const modifiedIdIncrement = (index: ModifiedId) =>
  `${baseId(index)}#${modifiedIdGetVersion(index) + 1}`;
