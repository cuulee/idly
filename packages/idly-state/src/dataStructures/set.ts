import { Either, Left, Right } from 'monet';

export const setAddIterable = <T>(
  it: IterableIterator<T> | Iterable<T>,
  set: Set<T>
) => {
  for (const i of it) {
    set.add(i);
  }
  return set;
};

export const setFind = <T>(
  foo: ((s: T) => boolean),
  set: Set<T> | ReadonlySet<T>
) => {
  for (const z of set) {
    if (foo(z)) {
      return z;
    }
  }
  return;
};

export const setEqual = <T>(setA: Set<T>, setB: Set<T>) => {
  if (setA.size !== setB.size) {
    return false;
  }
  for (const a of setA) {
    if (!setB.has(a)) {
      return false;
    }
  }
  return true;
};

export const setSome = <T>(
  foo: ((s: T) => boolean),
  set: Set<T> | ReadonlySet<T>
) => {
  for (const z of set) {
    if (foo(z)) {
      return true;
    }
  }
  return false;
};

export const setFilter = <T>(
  foo: ((s: T) => boolean),
  a: Set<T> | ReadonlySet<T>
) => new Set([...a].filter(foo));

export const setIntersect = <T>(
  a: Set<T> | ReadonlySet<T> | ReadonlySet<T>,
  b: Set<T> | ReadonlySet<T>
) => setFilter(x => b.has(x), a);

export const setDifference = <T>(
  a: Set<T> | ReadonlySet<T>,
  b: Set<T> | ReadonlySet<T>
) => new Set([...a].filter(x => !b.has(x)));

export const setCreate = <T>(a: Iterable<T> = []) => new Set<T>(a);

export const setBlankCreate = <T>() => new Set<T>();

export const setClone = <T>(s: ReadonlySet<T>) => new Set<T>([...s]);

export const setBulkDelete = <T>(
  a: Iterable<T> | IterableIterator<T>,
  set: Set<T>
) => {
  for (const i of a) {
    set.delete(i);
  }
  return set;
};
export const readonlySetCreate = <T>(a: Iterable<T> = []): ReadonlySet<T> =>
  new Set<T>(a);

export const iterableFlattenToArray = <T>(
  a: Iterable<Iterable<T>> | IterableIterator<IterableIterator<T>>
) => {
  const result = [];
  for (const i of a) {
    for (const j of i) {
      result.push(j);
    }
  }
  return result;
};

export const iterableFlattenToSet = <T>(
  a: Iterable<Iterable<T>> | IterableIterator<IterableIterator<T>>
): Set<T> => {
  const result = setCreate<T>();
  for (const i of a) {
    for (const j of i) {
      result.add(j);
    }
  }
  return result;
};

export const setUnion = <T>(
  a: Set<T> | ReadonlySet<T>,
  b: Set<T> | ReadonlySet<T>
) => new Set([...a, ...b]);

export const mapFilterKeys = <K, V>(
  foo: ((v: V, k: K) => boolean),
  map: Map<K, V>
): K[] => {
  const filter = [];
  for (const [k, v] of map) {
    if (foo(v, k)) {
      filter.push(k);
    }
  }
  return filter;
};

export const mapFilterKeysIntoSet = <K, V>(
  foo: ((v: V, k: K) => boolean),
  map: Map<K, V>
): Set<K> => {
  const filter = new Set();
  for (const [k, v] of map) {
    if (foo(v, k)) {
      filter.add(k);
    }
  }
  return filter;
};

export const mapFilter = <K, V>(
  foo: ((v: V, k: K) => boolean),
  map: Map<K, V>
): Map<K, V> => {
  const newMap = new Map<K, V>();
  for (const [k, v] of map) {
    if (foo(v, k)) {
      newMap.set(k, v);
    }
  }
  return newMap;
};

export const mapFilterValuesIntoArray = <K, V>(
  foo: ((v: V, k: K) => boolean),
  map: Map<K, V>
): V[] => {
  const filter = [];
  for (const [k, v] of map) {
    if (foo(v, k)) {
      filter.push(v);
    }
  }
  return filter;
};

export const last = <T>(r: T[]) => r[r.length - 1];

export const isVirgin = (indexOrId: string) => indexOrId.indexOf('#') === -1;
export const isNotVirgin = (indexOrId: string) => !isVirgin(indexOrId);

export const EitherCond = <T, K>(cond: boolean, left: T, right: K) =>
  cond ? Left<T, K>(left) : Right<T, K>(right);

/**
 * Returns Left(value) aka Error when fn(value) returns truthy else returns Right(error)
 */
export const ErrorWhen = <T, K>(fn: ((r: K) => boolean), error: T) => (
  value: K
) => (fn(value) ? Left<T, K>(error) : Right<T, K>(value));

export class Err {
  public static left<T = never>(err: Error): Either<Error, T> {
    return Left<Error, T>(err);
  }
  public static right<T>(val: T): Either<Error, T> {
    return Right(val);
  }
}

export const foldEitherArray = <T, K>(arr: Array<Either<T, K>>) => {
  const result: K[] = [];
  for (const x of arr) {
    if (x.isLeft()) {
      return Left<T, K[]>(x.left());
    }
    result.push(x.right());
  }
  return Right<T, K[]>(result);
};

export const pass = <T>(f: (r: T) => any) => (r: T) => {
  f(r);
  return r;
};
