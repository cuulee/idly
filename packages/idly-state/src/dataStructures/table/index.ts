export {
  OneToManyTable,
  oneToManyTableInsert,
  oneToManyTableUpdateIndex,
  oneToManyTableCreate,
  oneToManyTableFilter,
  oneToManyTableRemove,
  oneToManyTableRemoveIndex,
} from './oneToMany';

export {
  Table,
  tableAdd,
  tableUpdate,
  tableBulkAdd,
  tableBulkGet,
  tableBulkRemove,
  tableCreate,
  tableFilter,
  tableFindValue,
  tableGet,
  tableHas,
  tableMap,
  tableRemove,
  tableReduce,
  tableFindKey,
  tableForEach,
  tableCopy,
  ReadonlyTable,
} from './regular';
