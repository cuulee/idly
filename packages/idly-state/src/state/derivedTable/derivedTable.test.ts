import {
  nodeFactory,
  relationFactory,
  wayFactory,
} from 'idly-common/lib/osm/entityFactory';

import { setCreate } from '../../dataStructures/set';
import { derivedTableUpdate } from './index';

const n1 = nodeFactory({
  id: 'n1',
});
const n2 = nodeFactory({
  id: 'n2',
});
const n3 = nodeFactory({
  id: 'n3',
});
const n4 = nodeFactory({
  id: 'n4',
});
const n5 = nodeFactory({
  id: 'n5',
});

const n6 = nodeFactory({
  id: 'n6',
});

const n7 = nodeFactory({
  id: 'n7',
});

const w1 = wayFactory({
  id: 'w1',
  nodes: ['n1', 'n2'],
});

const w2 = wayFactory({
  id: 'w2',
  nodes: ['n1', 'n3'],
});

const w3 = wayFactory({
  id: 'w3',
  nodes: ['n3', 'n4', 'n5'],
});

const r1 = relationFactory({
  id: 'r1',
  members: [{ id: 'n4', ref: 'n4' }, { id: 'w2', ref: 'w2' }],
});

const r2 = relationFactory({
  id: 'r2',
  members: [
    { id: 'w1', ref: 'w1' },
    { id: 'w3', ref: 'w3' },
    { id: 'n6', ref: 'n6' },
  ],
});

const r3 = relationFactory({
  id: 'r3',
  members: [
    { id: 'r1', ref: 'r1' },
    { id: 'w2', ref: 'w2' },
    { id: 'n1', ref: 'n1' },
  ],
});

const mapFromObj = (o: any): Map<string, any> =>
  Object.keys(o).reduce((prev, k) => {
    prev.set(k, o[k]);
    return prev;
  }, new Map());

describe('derived values', () => {
  it('it throws error when way node doesnt exist', () => {
    const entityTable = mapFromObj({
      n1,
      r1,
      w1,
    });
    expect(() => derivedTableUpdate(entityTable)).toThrowError();
    // expect(derived.get('n1').parentWays).toEqual(setCreate(['w1']));

    // expect(derivedTableUpdate(entityTable)).toMatchSnapshot();
  });

  it('it handles relations whose members dont exist', () => {
    const entityTable = mapFromObj({
      n1,
      r1,
    });
    expect(() => derivedTableUpdate(entityTable)).not.toThrowError();
    // expect(derived.get('n1').parentWays).toEqual(setCreate(['w1']));

    // expect(derivedTableUpdate(entityTable)).toMatchSnapshot();
  });

  it('should mark parents correctly', () => {
    const entityTable = mapFromObj({
      n1,
      n2,
      n3,
      n4,
      n5,
      n6,
      n7,
      r1,
      r2,
      r3,
      w1,
      w2,
      w3,
    });
    const derived = derivedTableUpdate(entityTable);

    const parentWays: any = {};
    const parentRelations: any = {};

    derived.forEach(c => {
      parentWays[c.entity.id] = c.parentWays;
      parentRelations[c.entity.id] = c.parentRelations;
    });

    expect(parentWays).toMatchSnapshot();
    expect(parentRelations).toMatchSnapshot();
  });

  it('should remove entity if not found in the new table ', () => {
    let entityTable: any = mapFromObj({
      n1,
      n2,
      n3,
      n5,
      r1,
      w2,
    });

    const derived: any = derivedTableUpdate(entityTable);

    expect(derived.get('n1').parentWays).toEqual(setCreate(['w2']));

    entityTable = mapFromObj({
      n1,
      n3,
      n5,
      r1,
      w2,
    });

    const derived2 = derivedTableUpdate(entityTable, derived);
    expect(derived2.get('n1')).toBe(derived.get('n1'));
    expect(derived2.get('n5')).toBe(derived.get('n5'));

    expect(derived2.get('n2')).toBe(undefined);
  });

  it('should update the previous derived table if derived value changes', () => {
    let entityTable = mapFromObj({
      n1,
      n3,
      n4,
      n5,
      r1,
      w2,
      w3,
    });

    const baseDerivedTable = derivedTableUpdate(entityTable);
    expect(baseDerivedTable.get('n5')).toEqual({
      entity: n5,
      parentRelations: setCreate([]),
      parentWays: setCreate(['w3']),
    });

    entityTable = mapFromObj({
      n1,
      n3,
      n5,
      r1,
      w2,
    });

    const derived2 = derivedTableUpdate(entityTable, baseDerivedTable);

    expect(derived2.get('n5')).toEqual(baseDerivedTable.get('n5'));
    expect(derived2.get('n5')).toEqual({
      entity: n5,
      parentRelations: setCreate([]),
      parentWays: setCreate([]),
    });

    const newRelation = relationFactory({
      id: 'r55',
      members: [n1, n5, r1, w2, w3].map(e => ({ id: e.id, ref: e.id })),
    });

    const derived3 = derivedTableUpdate(
      mapFromObj({
        n1,
        n3,
        n4,
        n5,
        r1,
        r55: newRelation,
        w2,
        w3,
      }),
      baseDerivedTable
    );

    const parentWays: any = {};
    const parentRelations: any = {};

    derived3.forEach(c => {
      parentWays[c.entity.id] = c.parentWays;
      parentRelations[c.entity.id] = c.parentRelations;
    });

    expect(parentWays).toMatchSnapshot();
    expect(parentRelations).toMatchSnapshot();
    expect(baseDerivedTable).toMatchSnapshot();

    const derived5 = derivedTableUpdate(mapFromObj({}), baseDerivedTable);

    expect(derived5.size).toBe(0);
  });

  it('keeps the reference same when not modified', () => {
    let entityTable = mapFromObj({
      n1,
      n2,
      n3,
      n4,
      n5,
      n6,
      n7,
      r1,
      r2,
      r3,
      w1,
      w2,
      w3,
    });
    const derived = derivedTableUpdate(entityTable);

    entityTable = mapFromObj({
      n1,
      n2,
      n3,
      n4,
      n5,
      n6,
      n7,
      r1,
      r2,
      r3,
      w2,
      w3,
    });

    const derived2 = derivedTableUpdate(entityTable, derived);

    expect((derived2.get('n1') as any).parentWays).toEqual(setCreate(['w2']));
    expect((derived2.get('n2') as any).parentWays).toEqual(setCreate());

    expect(derived2.get('n3') as any).toBe(derived.get('n3') as any);
    expect(derived2.get('n4') as any).toBe(derived.get('n4') as any);
    expect(derived2.get('n5') as any).toBe(derived.get('n5') as any);
    expect(derived2.get('n6') as any).toBe(derived.get('n6') as any);
    expect(derived2.get('n7') as any).toBe(derived.get('n7') as any);
    expect(derived2.get('r1') as any).toBe(derived.get('r1') as any);
    expect(derived2.get('r2') as any).toBe(derived.get('r2') as any);
    expect(derived2.get('r3') as any).toBe(derived.get('r3') as any);
    expect(derived2.get('w2') as any).toBe(derived.get('w2') as any);
    expect(derived2.get('w3') as any).toBe(derived.get('w3') as any);

    expect((derived2.get('n3') as any).parentWays).toBe(
      (derived.get('n3') as any).parentWays
    );
    expect((derived2.get('n4') as any).parentWays).toBe(
      (derived.get('n4') as any).parentWays
    );
    expect((derived2.get('n5') as any).parentWays).toBe(
      (derived.get('n5') as any).parentWays
    );
    expect((derived2.get('n6') as any).parentWays).toBe(
      (derived.get('n6') as any).parentWays
    );
    expect((derived2.get('n7') as any).parentWays).toBe(
      (derived.get('n7') as any).parentWays
    );
    expect((derived2.get('r1') as any).parentWays).toBe(
      (derived.get('r1') as any).parentWays
    );
    expect((derived2.get('r2') as any).parentWays).toBe(
      (derived.get('r2') as any).parentWays
    );
    expect((derived2.get('r3') as any).parentWays).toBe(
      (derived.get('r3') as any).parentWays
    );
    expect((derived2.get('w2') as any).parentWays).toBe(
      (derived.get('w2') as any).parentWays
    );
    expect((derived2.get('w3') as any).parentWays).toBe(
      (derived.get('w3') as any).parentWays
    );
  });

  describe('should update the previous derived table if derived value changes 2', () => {
    let entityTable = mapFromObj({
      n1,
      n2,
      n3,
      n4,
      n5,
      n6,
      n7,
      r1,
      r2,
      r3,
      w1,
      w2,
      w3,
    });
    const derivedTable = derivedTableUpdate(entityTable);

    const w1Ref = derivedTable.get('w1');
    const r1Ref = derivedTable.get('r1');

    const n1Ref = derivedTable.get('n1');
    const n3Ref = derivedTable.get('n3');
    const n4Ref = derivedTable.get('n4');
    const n5Ref = derivedTable.get('n5');

    entityTable = mapFromObj({
      n1,
    });

    const compute1 = derivedTableUpdate(entityTable, derivedTable);

    const n1Ref2 = compute1.get('n1');
    expect(compute1.size).toBe(1);
    expect(compute1.get('n1')).toMatchSnapshot();

    it('updates the derived table when n1 is modified', () => {
      expect(derivedTable.get('n1')).not.toBe(n1Ref);
    });

    it('keeps other elements in derived table intact', () => {
      expect(derivedTable.get('w1')).toBe(w1Ref);
      expect(derivedTable.get('n3')).toBe(n3Ref);
    });

    derivedTableUpdate(
      mapFromObj({
        n1,
        n2,
        n3,
        n4,
        n5,
        n6,
        n7,
        r1,
        r2,
        r3,
        w1,
        w2,
        w3,
      }),
      derivedTable
    );
    it('changes the n1 ref ', () => {
      expect(derivedTable.get('n1')).not.toBe(n1Ref);
      expect(derivedTable.get('n1')).not.toBe(n1Ref2);
    });

    it('reverts n1 value back to the original', () => {
      expect(derivedTable.get('n1')).not.toEqual(n1Ref2);
      expect(derivedTable.get('n1')).toEqual(n1Ref);
    });

    it('keeps other elements in derived table intact', () => {
      expect(derivedTable.get('w1')).toBe(w1Ref);
      expect(derivedTable.get('n3')).toBe(n3Ref);
      expect(derivedTable.get('n4')).toBe(n4Ref);
      expect(derivedTable.get('n5')).toBe(n5Ref);
      expect(derivedTable.get('r1')).toBe(r1Ref);
    });
  });
});
