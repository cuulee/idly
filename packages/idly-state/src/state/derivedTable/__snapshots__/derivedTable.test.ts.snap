// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[` 1`] = `
Object {
  "entity": Object {
    "attributes": Object {
      "changeset": undefined,
      "timestamp": undefined,
      "uid": undefined,
      "user": undefined,
      "version": undefined,
      "visible": undefined,
    },
    "id": "n1",
    "loc": Object {
      "lat": 0,
      "lon": 0,
    },
    "tags": Object {},
    "type": "node",
  },
  "parentRelations": Set {},
  "parentWays": Set {},
}
`;

exports[`derived values should mark parents correctly 1`] = `
Object {
  "n1": Set {
    "w1",
    "w2",
  },
  "n2": Set {
    "w1",
  },
  "n3": Set {
    "w2",
    "w3",
  },
  "n4": Set {
    "w3",
  },
  "n5": Set {
    "w3",
  },
  "n6": Set {},
  "n7": Set {},
  "r1": Set {},
  "r2": Set {},
  "r3": Set {},
  "w1": Set {},
  "w2": Set {},
  "w3": Set {},
}
`;

exports[`derived values should mark parents correctly 2`] = `
Object {
  "n1": Set {
    "r3",
  },
  "n2": Set {},
  "n3": Set {},
  "n4": Set {
    "r1",
  },
  "n5": Set {},
  "n6": Set {
    "r2",
  },
  "n7": Set {},
  "r1": Set {
    "r3",
  },
  "r2": Set {},
  "r3": Set {},
  "w1": Set {
    "r2",
  },
  "w2": Set {
    "r1",
    "r3",
  },
  "w3": Set {
    "r2",
  },
}
`;

exports[`derived values should update the previous derived table if derived value changes 1`] = `
Object {
  "n1": Set {
    "w2",
  },
  "n3": Set {
    "w2",
    "w3",
  },
  "n4": Set {
    "w3",
  },
  "n5": Set {
    "w3",
  },
  "r1": Set {},
  "r55": Set {},
  "w2": Set {},
  "w3": Set {},
}
`;

exports[`derived values should update the previous derived table if derived value changes 2`] = `
Object {
  "n1": Set {
    "r55",
  },
  "n3": Set {},
  "n4": Set {
    "r1",
  },
  "n5": Set {
    "r55",
  },
  "r1": Set {
    "r55",
  },
  "r55": Set {},
  "w2": Set {
    "r1",
    "r55",
  },
  "w3": Set {
    "r55",
  },
}
`;

exports[`derived values should update the previous derived table if derived value changes 3`] = `
Map {
  "n1" => Object {
    "entity": Object {
      "attributes": Object {
        "changeset": undefined,
        "timestamp": undefined,
        "uid": undefined,
        "user": undefined,
        "version": undefined,
        "visible": undefined,
      },
      "id": "n1",
      "loc": Object {
        "lat": 0,
        "lon": 0,
      },
      "tags": Object {},
      "type": "node",
    },
    "parentRelations": Set {
      "r55",
    },
    "parentWays": Set {
      "w2",
    },
  },
  "n3" => Object {
    "entity": Object {
      "attributes": Object {
        "changeset": undefined,
        "timestamp": undefined,
        "uid": undefined,
        "user": undefined,
        "version": undefined,
        "visible": undefined,
      },
      "id": "n3",
      "loc": Object {
        "lat": 0,
        "lon": 0,
      },
      "tags": Object {},
      "type": "node",
    },
    "parentRelations": Set {},
    "parentWays": Set {
      "w2",
      "w3",
    },
  },
  "n4" => Object {
    "entity": Object {
      "attributes": Object {
        "changeset": undefined,
        "timestamp": undefined,
        "uid": undefined,
        "user": undefined,
        "version": undefined,
        "visible": undefined,
      },
      "id": "n4",
      "loc": Object {
        "lat": 0,
        "lon": 0,
      },
      "tags": Object {},
      "type": "node",
    },
    "parentRelations": Set {
      "r1",
    },
    "parentWays": Set {
      "w3",
    },
  },
  "n5" => Object {
    "entity": Object {
      "attributes": Object {
        "changeset": undefined,
        "timestamp": undefined,
        "uid": undefined,
        "user": undefined,
        "version": undefined,
        "visible": undefined,
      },
      "id": "n5",
      "loc": Object {
        "lat": 0,
        "lon": 0,
      },
      "tags": Object {},
      "type": "node",
    },
    "parentRelations": Set {
      "r55",
    },
    "parentWays": Set {
      "w3",
    },
  },
  "r1" => Object {
    "entity": Object {
      "attributes": Object {
        "changeset": undefined,
        "timestamp": undefined,
        "uid": undefined,
        "user": undefined,
        "version": undefined,
        "visible": undefined,
      },
      "id": "r1",
      "members": Array [
        Object {
          "id": "n4",
          "ref": "n4",
        },
        Object {
          "id": "w2",
          "ref": "w2",
        },
      ],
      "tags": Object {},
      "type": "relation",
    },
    "parentRelations": Set {
      "r55",
    },
    "parentWays": Set {},
  },
  "w2" => Object {
    "entity": Object {
      "attributes": Object {
        "changeset": undefined,
        "timestamp": undefined,
        "uid": undefined,
        "user": undefined,
        "version": undefined,
        "visible": undefined,
      },
      "id": "w2",
      "nodes": Array [
        "n1",
        "n3",
      ],
      "tags": Object {},
      "type": "way",
    },
    "parentRelations": Set {
      "r1",
      "r55",
    },
    "parentWays": Set {},
  },
  "w3" => Object {
    "entity": Object {
      "attributes": Object {
        "changeset": undefined,
        "timestamp": undefined,
        "uid": undefined,
        "user": undefined,
        "version": undefined,
        "visible": undefined,
      },
      "id": "w3",
      "nodes": Array [
        "n3",
        "n4",
        "n5",
      ],
      "tags": Object {},
      "type": "way",
    },
    "parentRelations": Set {
      "r55",
    },
    "parentWays": Set {},
  },
  "r55" => Object {
    "entity": Object {
      "attributes": Object {
        "changeset": undefined,
        "timestamp": undefined,
        "uid": undefined,
        "user": undefined,
        "version": undefined,
        "visible": undefined,
      },
      "id": "r55",
      "members": Array [
        Object {
          "id": "n1",
          "ref": "n1",
        },
        Object {
          "id": "n5",
          "ref": "n5",
        },
        Object {
          "id": "r1",
          "ref": "r1",
        },
        Object {
          "id": "w2",
          "ref": "w2",
        },
        Object {
          "id": "w3",
          "ref": "w3",
        },
      ],
      "tags": Object {},
      "type": "relation",
    },
    "parentRelations": Set {},
    "parentWays": Set {},
  },
}
`;
